server/src/main/java/org/elasticsearch/common/util/SetBackedScalingCuckooFilter.java

```diff
@@ -0,0 +1,409 @@
+        // Cache the chosen numBuckets for later use
+        numBuckets = t.getNumBuckets();
+        fingerprintMask = t.getFingerprintMask();
+        bitsPerEntry = t.getBitsPerEntry();
+
+        hashes.forEach(t::add);
+        filters.add(t);
+
+        hashes = null;
+        isSetMode = false;
+
+        breaker.accept(-oldSize); // this zeros out the overhead of the set
+        breaker.accept(getSizeInBytes()); // this adds back in the new overhead of the cuckoo filters
+
+    }
+
+    /**
+     * Get the approximate size of this datastructure.  Approximate because only the Set occupants
+     * are tracked, not the overhead of the Set itself.
+     */
+    public long getSizeInBytes() {
+        long bytes = 13; // fpp (double), threshold (int), isSetMode (boolean)
+        if (hashes != null) {
+            bytes = (hashes.size() * 16);
+        }
+        if (filters != null) {
+            bytes += filters.stream().mapToLong(CuckooFilter::getSizeInBytes).sum();
+        }
+        return bytes;
+    }
+
+
+    /**
+     * Merge `other` cuckoo filter into this cuckoo.  After merging, this filter's state will
+     * be the union of the two.  During the merging process, the internal Set may be upgraded
+     * to a cuckoo if it goes over threshold
+     */
+    public void merge(SetBackedScalingCuckooFilter other) {
+        // Some basic sanity checks to make sure we can merge
+        if (this.threshold != other.threshold) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because thresholds do not match: ["
+                + this.threshold + "] vs [" + other.threshold + "]");
+        }
+        if (this.capacity != other.capacity) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because capacities do not match: ["
+                + this.capacity + "] vs [" + other.capacity + "]");
+        }
+        if (this.fpp != other.fpp) {
+            throw new IllegalStateException("Cannot merge other CuckooFilter because precisions do not match: ["
+                + this.fpp + "] vs [" + other.fpp + "]");
+        }
+
+        if (isSetMode && other.isSetMode) {
+            // Both in sets, merge collections then see if we need to convert to cuckoo
+            hashes.addAll(other.hashes);
+            maybeConvert();
+        } else if (isSetMode && other.isSetMode == false) {
+            // Other is in cuckoo mode, so we convert our set to a cuckoo, then
+            // call the merge function again.  Since both are now in set-mode
+            // this will fall through to the last conditional and do a cuckoo-cuckoo merge
+            convert();
+            merge(other);
+        } else if (isSetMode == false && other.isSetMode) {
+            // Rather than converting the other to a cuckoo first, we can just
+            // replay the values directly into our filter.
+            other.hashes.forEach(this::add);
+            maybeConvert();
```

Comment:
Is maybeConvert() necessary?

Reply:
Ah no, good catch.  I don't think this is needed.